beacon-chain/sync/execution_payload_envelope.go

```diff
@@ -0,0 +1,102 @@
+package sync
+
+import (
+	"context"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/verification"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/interfaces"
+	"github.com/prysmaticlabs/prysm/v5/monitoring/tracing"
+	v1 "github.com/prysmaticlabs/prysm/v5/proto/engine/v1"
+	"go.opencensus.io/trace"
+	"google.golang.org/protobuf/proto"
+)
+
+func (s *Service) validateExecutionPayloadEnvelope(ctx context.Context, pid peer.ID, msg *pubsub.Message) (pubsub.ValidationResult, error) {
+	if pid == s.cfg.p2p.PeerID() {
+		return pubsub.ValidationAccept, nil
+	}
+	if s.cfg.initialSync.Syncing() {
+		return pubsub.ValidationIgnore, nil
+	}
+	ctx, span := trace.StartSpan(ctx, "sync.validateExecutionPayloadEnvelope")
+	defer span.End()
+	if msg.Topic == nil {
+		return pubsub.ValidationReject, errInvalidTopic
+	}
+	m, err := s.decodePubsubMessage(msg)
+	if err != nil {
+		tracing.AnnotateError(span, err)
+		return pubsub.ValidationReject, err
+	}
+	signedEnvelope, ok := m.(*v1.SignedExecutionPayloadEnvelope)
+	if !ok {
+		return pubsub.ValidationReject, errWrongMessage
+	}
+	e, err := blocks.WrappedROSignedExecutionPayloadEnvelope(signedEnvelope)
+	if err != nil {
+		log.WithError(err).Error("failed to create read only signed payload execution envelope")
+		return pubsub.ValidationIgnore, err
+	}
+	v := s.newExecutionPayloadEnvelopeVerifier(e, verification.GossipExecutionPayloadEnvelopeRequirements)
+
+	if err := v.VerifyBlockRootSeen(s.seenBlockRoot); err != nil {
+		return pubsub.ValidationIgnore, err
```

Comment:
If there's an invalid block, we want to REJECT here, not IGNORE. But `seenBlockRoot` will return true if the block root exists in bad block cache

Reply:
I was wrong. Please Ignore this. If `seenBlockRoot` returns true, we return nil and will not return pubsub here

Reply:
What are the criteria for deciding whether to return REJECT or IGNORE?

Reply:
REJECT downscores the peer, indicating an objective fault such as invalid messages, making it clear that the peer is bad. In contrast, IGNORE does not downscore the peer and is used for more subjective issues, like a late message. You can find more details here: https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#extended-validators

Reply:
In this case, `VerifyBlockRootSeen` returns non-nil error when `BeaconBlockRoot` is a bad block or not in fork choice. I'm trying to understand why a bad block is ignored instead of getting rejected. Is there a good definition of "bad" block?

Reply:
The bad block gets rejected in the next condition

Reply:
Ah right. Thank you for the answers!